### PR TITLE
Support setting properties for storage_root_path

### PR DESCRIPTION
**Pull request by vagetablechicken**
ref https://github.com/apache/incubator-doris/issues/2195

We can specify the properties of storage_root_path by setting '<property>:<value>', seperate by ','
e.g.
`storage_root_path = /home/disk1/palo,medium:ssd,capacity:50`

### CODE REVIEW

File: be/src/olap/options.h [hunk 1]
```diff
@@ -26,11 +26,14 @@
 namespace doris {
 
 struct StorePath {
-    StorePath() : capacity_bytes(-1) { }
+    StorePath() : capacity_bytes(-1), storage_medium(TStorageMedium::HDD){ }
```
Comment (imay) [inline be/src/olap/options.h]:
```suggestion
    StorePath() : capacity_bytes(-1), storage_medium(TStorageMedium::HDD) { }
```

File: be/src/olap/options.h [hunk 2]
```diff
@@ -26,11 +26,14 @@
...
     StorePath(const std::string& path_, int64_t capacity_bytes_)
         : path(path_), capacity_bytes(capacity_bytes_) { }
```
Comment (imay) [inline be/src/olap/options.h]: missing medium's default value?

File: be/src/olap/options.cpp [hunk 3]
```diff
@@ -47,18 +47,63 @@ OLAPStatus parse_conf_store_paths(
                 return OLAP_ERR_INPUT_PARAMETER_ERROR;
             }
 
-            // parse root path capacity
+            // parse root path capacity and storage medium
+            std::string capacity_str, medium_str;
```
Comment (imay) [inline be/src/olap/options.cpp]: Can you add some unit test for this code?

File: be/src/olap/options.cpp [hunk 4]
```diff
@@ -47,18 +47,63 @@ OLAPStatus parse_conf_store_paths(
...
+                    property = boost::trim_copy(tmp_vec[i].substr(0, found));
+                    value = boost::trim_copy(tmp_vec[i].substr(found + 1));
+                    LOG(INFO) << "get setting: "<< property <<": "<<value;
+                } else {
+                    // <value> only supports setting capacity
+                    property = "capacity";
+                    value = boost::trim_copy(tmp_vec[i]);
+                }
+
+                if (property == "capacity") {
```
Comment (imay) [inline be/src/olap/options.cpp]: better to use ignore case compare

File: be/src/olap/options.cpp [hunk 5]
```diff
@@ -47,18 +47,63 @@ OLAPStatus parse_conf_store_paths(
...
+            }
+
+            for (int i = 1; i < tmp_vec.size(); i++) {
+                // <property>:<value> or <value>
+                std::string property, value;
+                std::size_t found = tmp_vec[i].find(':');
+                if (found != std::string::npos) {
+                    property = boost::trim_copy(tmp_vec[i].substr(0, found));
+                    value = boost::trim_copy(tmp_vec[i].substr(found + 1));
+                    LOG(INFO) << "get setting: "<< property <<": "<<value;
```
Comment (imay) [inline be/src/olap/options.cpp]: remove this log

**Comment by vagetablechicken** [comment]
property medium has a higher priority, so I decide to set storage_medium in c'tor of DataDir, change 
'_init_extension_and_capacity()' to '_init_capacity()'. 
https://github.com/apache/incubator-doris/blob/14769b0beb7671626da011611d2c73ffe841b831/be/src/olap/data_dir.h#L132

But in meta_tool, 
https://github.com/apache/incubator-doris/blob/14769b0beb7671626da011611d2c73ffe841b831/be/src/tools/meta_tool.cpp#L143
var 'data_dir' will not be set the _storage_medium, cause I remove the _init_extension part.

I don't want to redefine TStorageMedium enum, e.g. {UNSET, HDD,SSD}.
How about 
```
    DataDir(const std::string& path,
            int64_t capacity_bytes = -1,
            std::string medium_property = "",
            TabletManager* tablet_manager = nullptr,
            TxnManager* txn_manager = nullptr);
```
Thus, 
1. medium_property != "", set _storage_medium
1. medium_property == "", use extension

File: be/src/olap/options.h (hunk 1 above)
Comment (vagetablechicken) [reply be/src/olap/options.h]: How do you format the code? I can't find .clang-format, is it a predefined style?

Comment (imay) [reply be/src/olap/options.h]: The code in Doris comes from many different projects. Generally, the style of each file is consistent, if you add code to a file, you should keep style with it. If it is a new file, Baidu's code style is used. It's about the same style as Google's. Except lowercase and underscore as the function name, and use 4-byte indentation and so on.

clang-format is a good suggestion, we should introduce it to our project to make developers easier.

**Comment by imay** [comment]
@vagetablechicken 

1. I think meta tool should not care about the medium of data path.
2. Even if meta tool care about medium, it is better to make medium and capacity parse in one place. You can make config parse a function, which can be called by both options and meta dir.

File: be/src/olap/options.h [hunk 6]
```diff
@@ -26,14 +26,25 @@
 namespace doris {
 
 struct StorePath {
-    StorePath() : capacity_bytes(-1) { }
-    StorePath(const std::string& path_, int64_t capacity_bytes_)
-        : path(path_), capacity_bytes(capacity_bytes_) { }
+    StorePath() : capacity_bytes(-1), storage_medium(TStorageMedium::HDD) { }
+    StorePath(const std::string &path_, int64_t capacity_bytes_)
```
Comment (imay) [inline be/src/olap/options.h]:
```suggestion
    StorePath(const std::string& path_, int64_t capacity_bytes_)
```
Doris prefer `const std::string& path` to `const std::string &path`

File: be/src/olap/options.cpp [hunk 7]
```diff
@@ -29,36 +29,96 @@ namespace doris {
...
-                if (!valid_signed_number<int64_t>(tmp_vec[1])
-                        || strtol(tmp_vec[1].c_str(), NULL, 10) < 0) {
-                    LOG(WARNING) << "invalid capacity of store path, capacity=" << tmp_vec[1];
-                    return OLAP_ERR_INPUT_PARAMETER_ERROR;
-                }
-                capacity_bytes = strtol(tmp_vec[1].c_str(), NULL, 10) * GB_EXCHANGE_BYTE;
-            }
-
-            paths->emplace_back(tmp_vec[0], capacity_bytes);
+            paths->emplace_back(path.path, path.capacity_bytes,
```
Comment (imay) [inline be/src/olap/options.cpp]:
```suggestion
            paths->emplace_back(std::move(path));
```

File: be/src/olap/options.h [hunk 8]
```diff
@@ -26,14 +26,25 @@
...
+        : path(path_),
+          capacity_bytes(capacity_bytes_),
+          storage_medium(storage_medium_) { }
     std::string path;
     int64_t capacity_bytes;
+    TStorageMedium::type storage_medium;
 };
 
-OLAPStatus parse_conf_store_paths(const std::string& config_path, std::vector<StorePath>* path);
+OLAPStatus parse_root_path(const std::string &root_path, StorePath *path);
```
Comment (imay) [inline be/src/olap/options.h]: better to add some comment for this function to let other known what this function does.
```suggestion
OLAPStatus parse_root_path(const std::string& root_path, StorePath* path);
```

File: be/src/olap/options.cpp [hunk 9]
```diff
@@ -29,36 +29,96 @@ namespace doris {
...
+    if (tmp_vec[0].empty() || tmp_vec[0][0] != '/') {
+        LOG(WARNING) << "invalid store path. path=" << tmp_vec[0];
+        return OLAP_ERR_INPUT_PARAMETER_ERROR;
+    }
+    path->path = tmp_vec[0];
+
+    // parse root path capacity and storage medium
+    std::string capacity_str, medium_str;
+
+    boost::filesystem::path boost_path = tmp_vec[0];
```
Comment (imay) [inline be/src/olap/options.cpp]: should do this operation in a try catch block to avoid exception

File: be/src/olap/options.cpp [hunk 10]
```diff
@@ -29,36 +29,96 @@ namespace doris {
...
+        std::size_t found = tmp_vec[i].find(':');
+        if (found != std::string::npos) {
+            property = boost::trim_copy(tmp_vec[i].substr(0, found));
+            value = boost::trim_copy(tmp_vec[i].substr(found + 1));
+        } else {
+            // <value> only supports setting capacity
+            property = "capacity";
+            value = boost::trim_copy(tmp_vec[i]);
+        }
+        if (property == "capacity") {
```
Comment (imay) [inline be/src/olap/options.cpp]:
```suggestion
        if (boost::iequals(property, "capacity")) {
```

File: be/src/olap/options.h (hunk 6 above)
Comment (vagetablechicken) [reply be/src/olap/options.h]: So Doris align pointer to the right, and reference to the left? I think that's not a good idea.
In clang-format, pointer and reference alignment style is a single option, "PointerAlignment".
http://clang.llvm.org/docs/ClangFormatStyleOptions.html

Comment (imay) [reply be/src/olap/options.h]: Doris align both reference and pointer to the left.

Comment (vagetablechicken) [reply be/src/olap/options.h]: ok, I chanced to see this, mistakely think that Doris align pointer to the right   
https://github.com/apache/incubator-doris/blob/42a4fff562090fc40bcab48684b4a61614f8d8f8/be/src/tools/meta_tool.cpp#L105

**Review by imay — APPROVED** [review]
Good Job
LGTM

**Pull request merged**